iam_module

juliedavila opened the pull request:
This is a resubmission of https://github.com/ansible/ansible-modules-core/pull/590 in order to have a clean commit history. 

There is currently no IAM module. This is part 1 of 3 that I plan to make.

Part 1: Create/Update/Delete users, groups and roles
Part 2: Create/Update/Delete policy objects for users/groups/roles See https://github.com/ansible/ansible-modules-core/pull/887
Part 3: Upload/Remove/Update server certificates. See #918 

I opted for compartmentalization to avoid entering ec2_vpc's monolithic state/complexity.

I've done a good amount of testing but def should get poked some more.

brian-brazil commented on cloud/amazon/iam.py:
```diff
@@ -0,0 +1,539 @@
+      - a list of the keys that you want impacted by the access_key_state paramter.
+  groups:
+    description:
+      - A list of groups the user should belong to. When update, will gracefully the remove groups not listed.
```
I think there's an extra 'the' here

juliedavila replied:
Maybe I'm going crazy but I dont see the extra 'the' which line specifically are you referring to?

brian-brazil replied:
"will gracefully the remove groups not listed" should probably be "will gracefully remove groups not listed"

juliedavila replied:
Ah! I was looking for 'the the'. Thanks. Fixed. Will update the P/R within the hour with this change and the stuff mentioned below.

bcoca commented:
several things:
- doc parsing has issues, mostly because all descriptions have to be lists ( start with -) and : is a special character and must be quoted or avoided.
- aws errors are not captured and exceptions are allowed to fall through, these should return in a fail_json call to the module.

```
failed: [localhost] => (item=test42) => {"failed": true, "item": "test42", "parsed": false}
Traceback (most recent call last):
  File "/home/bcoca/.ansible/tmp/ansible-tmp-1424109800.44-46630851071663/iam", line 2335, in <module>                                                                            
    main()                                                                                                                                                                        
  File "/home/bcoca/.ansible/tmp/ansible-tmp-1424109800.44-46630851071663/iam", line 426, in main                                                                                 
    iam, name, password, path, key_state)                                                                                                                                         
  File "/home/bcoca/.ansible/tmp/ansible-tmp-1424109800.44-46630851071663/iam", line 165, in create_user                                                                          
    pwd = iam.create_login_profile(name, pwd)                                                                                                                                     
  File "/usr/lib64/python2.7/site-packages/boto/iam/connection.py", line 954, in create_login_profile                                                                             
    return self.get_response('CreateLoginProfile', params)                                                                                                                        
  File "/usr/lib64/python2.7/site-packages/boto/iam/connection.py", line 102, in get_response                                                                                     
    raise self.ResponseError(response.status, response.reason, body)                                                                                                              
boto.exception.BotoServerError: BotoServerError: 400 Bad Request                                                                                                                  
<ErrorResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">                                                                                                                 
  <Error>                                                                                                                                                                         
    <Type>Sender</Type>                                                                                                                                                           
    <Code>PasswordPolicyViolation</Code>                                                                                                                                          
    <Message>Password does not conform to the account password policy.</Message>                                                                                                  
  </Error>                                                                                                                                                                        
  <RequestId>d391166e-b605-11e4-9256-31147537fd9a</RequestId>                                                                                                                     
</ErrorResponse>   
```

also there seems to be an issue with user_groups

```
failed: [localhost] => (item=test42) => {"failed": true, "item": "test42", "parsed": false}
Traceback (most recent call last):
  File "/home/bcoca/.ansible/tmp/ansible-tmp-1424109955.91-207947638917259/iam", line 2340, in <module>
    main()
  File "/home/bcoca/.ansible/tmp/ansible-tmp-1424109955.91-207947638917259/iam", line 447, in main
    user_meta=user, keys=keys, groups=user_groups, changed=changed)
UnboundLocalError: local variable 'user_groups' referenced before assignment
```

this seems mostly bad return when groups don't exist as it is outside the "if" that checks for groups

bcoca commented:
also, 'delete_all_keys' function is called but it is not defined anywhere

Commit:
iam module with fixes

iam module. fix policy issue.

bugfix

allow for modifying path without declaring new_path

juliedavila commented:
Patched up. Identified some other issues as well which have also been fixed.

juliedavila commented:
One last issue to address.  With boto, a role/user/group cannot be deleted until all policy objects have been detached from the resource. A bit of a challenge given that boto doesn't support detaching Managed policies. Would need to add a disclaimer in the docs.

bcoca commented:
ok, add disclaimer (note might work) and I'll be able to give this another look by EOW

juliedavila commented:
Updated.

rahulmehrotra1 commented:
@defionscode Can you please provide an example for iam role creation using the module in the documentation ?
I am not able to create a role using the following

> name: Creating eventlogger role
> local_action:
> module: iam
> iam_type: role
> name: eventlogger
> state: present

getting the following error

```
TASK: [Creating eventlogger role] ********************************************* 
failed: [localhost -> 127.0.0.1] => {"failed": true, "parsed": false}
Traceback (most recent call last):
  File "/home/rahul/.ansible/tmp/ansible-tmp-1426115965.92-241507198509407/iam", line 2499, in <module>
    main()
  File "/home/rahul/.ansible/tmp/ansible-tmp-1426115965.92-241507198509407/iam", line 694, in main
    iam, name, path, orig_role_list, orig_prof_list)
TypeError: create_role() takes exactly 6 arguments (5 given)
```

juliedavila commented:
@rhlmhrtr my fault. The way you used it should have created that role. I'll fix that now.

juliedavila commented:
@rhlmhrtr yep it was a one word fix. Your example should work now.

rahulmehrotra1 commented:
@defionscode awsm, just tried it out. Thanks

Constantin07 commented:
@defionscode 
The update doesn't work properly. How to reproduce:
1. create a user with path: /division1/
2. change the path to /division2/
3. run playbook once again. It reports "changed=0" but in AWS console the path is old one.
no matter what state is specified "present" or "update"

Constantin07 commented:
@defionscode 

```
- name: Test IAM
  iam:
   iam_type: role
   name: aaa
   state: present
```

Creation of role fails with such error:
TASK: [Query IAM] *************************************************************
<localhost> REMOTE_MODULE iam iam_type=role name=aaa state=present
failed: [localhost] => {"failed": true, "parsed": false}
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1427922366.45-15755177960783/iam", line 2325, in <module>
    main()
  File "/root/.ansible/tmp/ansible-tmp-1427922366.45-15755177960783/iam", line 379, in main
    key_state = module.params.get('access_key_state').lower()
AttributeError: 'NoneType' object has no attribute 'lower'

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/root/aws-iam.retry

localhost                  : ok=0    changed=0    unreachable=0    failed=1

PS: AWS access and secret keys are defined as variables and work fine with other AWS modules.

juliedavila commented:
@Constantin07 I am unable to replicate your role creation issue with your exact test case. Have you refetched this PR?

Checking the path change issue now.

juliedavila commented:
@Constantin07 I see the issue, the way I designed it was to require `new_path` if you wanted to change the path. This makes sense for the username, but for path I think I should probably refactor it so it can be done the way you attempted. I'll patch it.

juliedavila commented:
@Constantin07 I've added that functionality and tested. Now you can switch paths via `state=update` and `new_path` or just by specifying a path that doesn't match the user's current path.

bpesics commented:
@defionscode great stuff, but why don't you submit these modules instead of core to ansible-modules-extras first so that the barrier to get used and shipped is lower?

juliedavila commented:
@bpesics right now the team is set on waiting for v2 before merging new modules. Whether I placed it here or extras, the wait would be the same. Then there is the consistency bit of keeping AWS modules together.

sivel commented:
Despite the desire for consistency, _all_ new modules should be submitted to -extras as per the instructions in the README.md for this repo:

> New module submissions for modules that do not yet exist should be submitted to ansible-modules-extras, rather than this repo.

juliedavila commented:
@sivel an internal discussion took place, these particular modules (IAM) will go directly into this repo.

adamchainz commented:
I've just tried this out. Our codebase has a strict flake8 linter + 120 chars policy so I had to de-lint it. In doing so I found a couple unnecessary imports, unclear lines, etc. and also one apparent bug, which I'll report on the appropriate PR.

My de-linted _iam.py_ is at https://gist.github.com/adamchainz/7f5834fbbfa8e72c48da#file-iam-py

The only other change asides from flake8 compatibility I made was to make `state` default to `present`, like all other ansible modules.

bcoca commented:
actually we prefer that state is required and not to default to anything

adamchainz commented:
Okay, I just found it very surprising since every one of `apt`, `yum`, `ec2`, etc. defaults to `present`.

echiu64 commented:
Would be nice to have a way to manage IAM policies vs inline policies attached to a user/group/role.

Wondering if iam_policy should maybe be a key in iam vs on it's own, since it's not really a standalone policy but an inline policy, so may as well inline it into iam? Looks like it's a limitation in boto right now (https://github.com/boto/boto/issues/2956)

But moving it into iam paves way to have an "iam_policy" that manages standalone policy objects.

Edit: Looks like support will be there soon: https://github.com/boto/boto/pull/3185

juliedavila commented:
@echiu64 yep, agreed. Managed policies are also non supported in boto. Boto3 should be a godsend.

Constantin07 commented:
Hi,

When creating the iam role

```
  - name: Create IAM roles
    iam:
     iam_type: role
     name: "{{ item }}"
     state: present
    with_items:
      - webservers
      - apservers
```

 the below error is thrown if 'access_key_state' is omitted:

```
TASK: [Create IAM roles] ******************************************************
<localhost> REMOTE_MODULE iam iam_type=role name=webservers state=present
failed: [localhost] => (item=webservers) => {"failed": true, "item": "webservers", "parsed": false}
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1433171112.85-143211962364645/iam", line 2330, in <module>
    main()
  File "/root/.ansible/tmp/ansible-tmp-1433171112.85-143211962364645/iam", line 379, in main
    key_state = module.params.get('access_key_state').lower()
AttributeError: 'NoneType' object has no attribute 'lower'
```

The user has nothing to do with roles. I think this can fix the problem:

```
@@ -376,7 +376,8 @@ def main():
     password = module.params.get('password')
     path = module.params.get('path')
     new_path = module.params.get('new_path')
-    key_state = module.params.get('access_key_state').lower()
+    if module.params.get('access_key_state') is not None:
+        key_state = module.params.get('access_key_state').lower()
     key_ids = module.params.get('access_key_ids')

     if iam_type == 'user' and module.params.get('password') is not None:
```

Kind Regards,
C

juliedavila commented:
@Constantin07 

The current code I see is:

``` python
key_state = module.params.get('access_key_state')
    if key_state:
        key_state = key_state.lower()
        if any([n in key_state for n in ['active', 'inactive']]) and not key_ids:
            module.fail_json(changed=False, msg="At least one access key has to be defined in order"
                                                " to use 'active' or 'inactive'")
```

Which should only be attempting the `lower()` if the key_state is not None. Can you confirm that you have the latest PR code?

Constantin07 commented:
Thanks @defionscode. It turns out I used an outdated version of code.
Please ignore my previous message.

evanccnyc commented:
Any update on when this and iam_policy will be checked in?

bcoca commented:
@defionscode are all issues resolved?

juliedavila commented:
Waiting on @bennojoy to finish up with some testing. He found a bug when attempting to remove a user. It's been patched. Not sure if he's gotten around to iam_policy yet. AFAIK, all known issues have been squashed.

bcoca commented:
bennojoy gave his OK
